src/python/tests/gff3/test_extract_annotation.py

```diff
@@ -22,15 +22,230 @@
+            in_child_id: ID for the SeqFeature child.
+            out_parent_type: Type for the parent stored in the FunctionalAnnotation.
+            out_parent_id: ID for the parent stored.
+            expected: What exception is expected to be raised, if any.
+
+        """
+        annot = FunctionalAnnotations()
+        parent = SeqFeature(type=in_parent_type, id=in_parent_id)
+        annot.add_feature(parent, "gene")
+        annot.add_feature(
+            SeqFeature(type="mRNA", id=in_child_id), feat_type="transcript", parent_id=in_parent_id
+        )
+
+        with expected:
+            out_parent = annot.get_parent(out_parent_type, out_child_id)
+            assert out_parent == in_parent_id
+
+    @pytest.mark.parametrize(
+        "child_type, child_id, out_parent_id, expected",
+        [
+            ("transcript", "mrna_A", "gene_A", does_not_raise()),
+            ("bad_type", "mrna_A", "gene_A", raises(KeyError)),  # Child type does not exist
+            ("gene", "gene_A", None, raises(AnnotationError)),  # Feature ID already loaded
+            ("gene", "gene_B", "gene_A", raises(AnnotationError)),  # Cannot add a gene child of a gene
+        ],
+    )
+    def test_add_feature_failures(
+        self, child_type: str, child_id: str, out_parent_id: Optional[str], expected: ContextManager
+    ):
+        """Tests the `FunctionaAnnotation.add_feature()` method failures.
+
+        Test the addition of a child feature after a parent has already been added.
+
+        Args:
+            child_type: Type for the SeqFeature child.
+            child_id: ID for the SeqFeature child.
+            out_parent_id: ID for the parent.
+            expected: What exception is expected to be raised, if any.
+
+        """
+        annot = FunctionalAnnotations()
+        with expected:
+            parent = SeqFeature(type="gene", id="gene_A")
+
+            child = SeqFeature(type="mRNA", id=child_id)
+            annot.add_feature(parent, "gene")
+            annot.add_feature(child, child_type, out_parent_id)
+
+    @pytest.mark.parametrize(
+        "feat_type, expected",
+        [
+            ("gene", does_not_raise()),
+            ("transcript", does_not_raise()),
+            ("translation", does_not_raise()),
+            ("bad_type", raises(KeyError)),
+        ],
+    )
+    def test_get_features(self, feat_type: str, expected: ContextManager):
+        """Tests the `FunctionaAnnotation.get_features()` method.
+
+        Load 2 features, then test the fetching of those features.
+
+        Args:
+            feat_type: Type for the features to fetch.
+            expected: What exception is expected to be raised, if any.
+
+        """
+        annot = FunctionalAnnotations()
+        one_gene = SeqFeature(type="gene", id="gene_A")
+        one_transcript = SeqFeature(type="mRNA", id="mrna_A")
+        annot.add_feature(one_gene, "gene")
+        annot.add_feature(one_transcript, "transcript", parent_id=one_gene.id)
```

Comment:
Do we want to create a dependency so this test is not run if `add_feature()` tests don't pass?